Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -106,22 +120,96 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples,
   // Flush cached variable
   m_indexR.store(indexR);
 
-  return numSamples;
+  return actual_sample_count;
 }
 
-unsigned int CMixer::Mix(short* samples, unsigned int num_samples, bool consider_framelimit)
+unsigned int CMixer::Mix(short* samples, unsigned int num_samples)
 {
   if (!samples)
     return 0;
 
   memset(samples, 0, num_samples * 2 * sizeof(short));
 
-  m_dma_mixer.Mix(samples, num_samples, consider_framelimit);
-  m_streaming_mixer.Mix(samples, num_samples, consider_framelimit);
-  m_wiimote_speaker_mixer.Mix(samples, num_samples, consider_framelimit);
+  if (SConfig::GetInstance().m_audio_stretch)
+  {
+    unsigned int available_samples =
+        std::min(m_dma_mixer.AvailableSamples(), m_streaming_mixer.AvailableSamples());
+
+    m_dma_mixer.Mix(m_stretch_buffer.data(), available_samples, false);
+    m_streaming_mixer.Mix(m_stretch_buffer.data(), available_samples, false);
+    m_wiimote_speaker_mixer.Mix(m_stretch_buffer.data(), available_samples, false);
+
+    if (!m_is_stretching)
+    {
+      m_sound_touch.clear();
+      m_is_stretching = true;
+    }
+    StretchAudio(m_stretch_buffer.data(), available_samples, samples, num_samples);
+  }
+  else
+  {
+    m_dma_mixer.Mix(samples, num_samples, true);
+    m_streaming_mixer.Mix(samples, num_samples, true);
+    m_wiimote_speaker_mixer.Mix(samples, num_samples, true);
+    m_is_stretching = false;
+  }
+
   return num_samples;
 }
 
+void CMixer::StretchAudio(const short* in, unsigned int num_in, short* out, unsigned int num_out)
+{
+  const double time_delta = static_cast<double>(num_out) / m_sampleRate;  // seconds
+
+  // We were given actual_samples number of samples, and num_samples were requested from us.
+  double current_ratio = static_cast<double>(num_in) / static_cast<double>(num_out);
+
+  const double max_latency = SConfig::GetInstance().m_audio_stretch_max_latency;
+  const double max_backlog = m_sampleRate * max_latency / 1000.0;
+  const double backlog_fullness = m_sound_touch.numSamples() / max_backlog;
+  if (backlog_fullness > 1.0)
+  {
+    // Exceeded latency budget: Do not add more samples into FIFO.
+    num_in = 0;
+  }
+
+  // We ideally want the backlog to be about 50% full.
+  // This gives some headroom both ways to prevent underflow and overflow.
+  // We tweak current_ratio to encourage this.
+  constexpr double tweak_time_scale = 0.1;  // seconds
+  current_ratio *= (2.0 * backlog_fullness - 1.0) * (time_delta / tweak_time_scale);
```

Comment:
Also, 100 ms is not that much. Bad backends aren't that much better.
As this is a D control system, it has a resonance. Likely if num_out >= 100 ms (or times two?).

I think this is fine, but we should keep it in mind ;)